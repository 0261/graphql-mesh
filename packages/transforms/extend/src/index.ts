import { GraphQLSchema, extendSchema, parse } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';

export const extendTransform: TransformFn<YamlConfig.Extend> = ({
  schema,
  config
}): GraphQLSchema => {
  return extendSchema(schema, parse(config.sdl));
};

export default extendTransform;
