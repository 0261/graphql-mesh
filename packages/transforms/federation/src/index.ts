import { GraphQLSchema, GraphQLObjectType, GraphQLID, isNonNullType, GraphQLNonNull } from 'graphql';
import { Transform, YamlConfig, MeshTransformOptions } from '@graphql-mesh/types';
import { loadFromModuleExportExpressionSync } from '@graphql-mesh/utils';
import { transformSchemaFederation, FederationConfig, FederationFieldsConfig } from 'graphql-transform-federation';

export default class FederationTransform implements Transform {
  constructor(private options: MeshTransformOptions<YamlConfig.Transform['federation']>) {}
  transformSchema(schema: GraphQLSchema) {
    const { config } = this.options;
    const federationConfig: FederationConfig<any> = {};

    if (config?.types) {
      for (const type of config.types) {
        const fields: FederationFieldsConfig = {};
        if (type.config?.fields) {
          for (const field of type.config.fields) {
            fields[field.name] = field.config;
          }
        }
        // If a field is a key field, it should be GraphQLID

        if (type.config?.keyFields) {
          for (const fieldName of type.config.keyFields) {
            const objectType = schema.getType(type.name) as GraphQLObjectType;
            const existingType = objectType.getFields()[fieldName].type;
            objectType.getFields()[fieldName].type = isNonNullType(existingType)
              ? GraphQLNonNull(GraphQLID)
              : GraphQLID;
          }
        }

        const resolveReference =
          type.config?.resolveReference && loadFromModuleExportExpressionSync(type.config.resolveReference);
        federationConfig[type.name] = {
          ...type.config,
          resolveReference,
          fields,
        };
      }
    }

    return transformSchemaFederation(schema, federationConfig);
  }
}
