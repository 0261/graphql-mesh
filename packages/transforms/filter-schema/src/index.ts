import { matcher } from 'micromatch';

import { GraphQLSchema } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';
import { Transform } from '@graphql-tools/utils';
import { FilterRootFields, FilterObjectFields, wrapSchema } from '@graphql-tools/wrap';

const filterSchemaTransform: TransformFn<YamlConfig.Transform['filterSchema']> = ({
  schema,
  config,
}): GraphQLSchema => {
  const rootTypes = [
    schema.getQueryType()?.name,
    schema.getMutationType()?.name,
    schema.getSubscriptionType()?.name,
  ].filter(Boolean) as string[];
  const transforms: Transform[] = [];
  for (const filter of config) {
    const [typeName, fieldGlob] = filter.split('.');
    const isMatch = matcher(fieldGlob.trim());
    if (rootTypes.includes(typeName)) {
      transforms.push(
        new FilterRootFields((rootTypeName, rootFieldName) => {
          if (rootTypeName === typeName) {
            return isMatch(rootFieldName);
          }
          return true;
        })
      );
    } else {
      transforms.push(
        new FilterObjectFields((objectTypeName, objectFieldName) => {
          if (objectTypeName === typeName) {
            return isMatch(objectFieldName);
          }
          return true;
        })
      );
    }
  }
  return wrapSchema({
    schema,
    transforms,
  });
};

export default filterSchemaTransform;
