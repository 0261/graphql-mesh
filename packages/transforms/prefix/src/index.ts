import { GraphQLSchema } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';
import { SchemaComposer } from 'graphql-compose';

export const prefixTransform: TransformFn<YamlConfig.Prefix> = async ({
  apiName,
  schema,
  config
}): Promise<GraphQLSchema> => {
  let prefix: string | null = null;

  if (config.prefix) {
    prefix = config.prefix;
  } else if (apiName) {
    prefix = `${apiName}_`;
  }

  if (!prefix) {
    throw new Error(`Transform 'prefix' has missing config: prefix`);
  }

  const composer = new SchemaComposer(schema);
  const excluded: string[] = [
    schema.getQueryType()?.name,
    schema.getMutationType()?.name,
    schema.getSubscriptionType()?.name,
    ...(config.ignore || [])
  ].filter(Boolean) as string[];
  const typeMap = schema.getTypeMap();

  for (const typeName of Object.keys(typeMap)) {
    if (!excluded.includes(typeName) && !typeName.startsWith('__')) {
      composer.get(typeName).setTypeName(`${prefix}${typeName}`);
    }
  }

  return composer.buildSchema();
};

export default prefixTransform;
