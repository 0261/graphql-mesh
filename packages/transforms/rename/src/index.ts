import { GraphQLSchema } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';
import { SchemaComposer } from 'graphql-compose';

export const renameTransform: TransformFn<YamlConfig.Rename> = ({
  schema,
  config
}): GraphQLSchema => {
  const composer = new SchemaComposer(schema);
  composer.get(config.from).setTypeName(config.to);

  return composer.buildSchema();
};

export default renameTransform;
