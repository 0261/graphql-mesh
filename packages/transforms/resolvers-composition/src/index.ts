import { GraphQLSchema } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';
import { addResolversToSchema } from '@graphql-tools/schema';
import { composeResolvers, ResolversComposerMapping } from '@graphql-tools/resolvers-composition';
import { extractResolvers, loadFromModuleExportExpression } from '@graphql-mesh/utils';

const resolversCompositionTransform: TransformFn<YamlConfig.ResolversCompositionTransformObject[]> = async ({
  schema,
  config,
}): Promise<GraphQLSchema> => {
  const resolversComposition: ResolversComposerMapping = {};

  for (const { resolver, composer } of config) {
    resolversComposition[resolver] = await loadFromModuleExportExpression(composer, 'default');
  }

  const resolvers = extractResolvers(schema);
  const composedResolvers = composeResolvers(resolvers, resolversComposition);

  return addResolversToSchema({
    schema,
    resolvers: composedResolvers,
    updateResolversInPlace: true,
  });
};

export default resolversCompositionTransform;
