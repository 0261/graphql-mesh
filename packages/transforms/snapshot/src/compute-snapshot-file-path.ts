import { isAbsolute, join } from 'path';
import objectHash from 'object-hash';
import { GraphQLResolveInfo } from 'graphql';

export function computeSnapshotFilePath(options: {
  info: GraphQLResolveInfo;
  args: any;
  outputDir: string;
  respectSelectionSet?: boolean;
}) {
  const typeName = options.info.parentType.name;
  const fieldName = options.info.fieldName;
  const hashObj = options.respectSelectionSet
    ? {
        args: options.args,
        fieldNodes: options.info.fieldNodes,
      }
    : options.args;
  const hash = objectHash(hashObj, { ignoreUnknown: true });
  const fileName = [typeName, fieldName, hash].join('_') + '.json';
  const absoluteOutputDir = isAbsolute(options.outputDir) ? options.outputDir : join(process.cwd(), options.outputDir);
  return join(absoluteOutputDir, fileName);
}
