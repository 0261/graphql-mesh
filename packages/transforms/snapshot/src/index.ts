import { GraphQLSchema } from 'graphql';
import { TransformFn, YamlConfig } from '@graphql-mesh/types';
import { addResolversToSchema } from '@graphql-tools/schema';
import { composeResolvers, ResolversComposerMapping, ResolversComposition } from '@graphql-tools/resolvers-composition';
import { isAbsolute, join } from 'path';
import { ensureFile, writeJSON, pathExists, readJSON } from 'fs-extra';
import { computeSnapshotFilePath } from './compute-snapshot-file-path';
import { extractResolvers } from '@graphql-mesh/utils';

const writeFile = async (path: string, json: any) => {
  try {
    await ensureFile(path);
    await writeJSON(path, json, {
      spaces: 2,
    });
  } catch (e) {
    console.error(`Snapshot cannot saved to ${path}: ${e.message}`);
  }
};

const snapshotTransform: TransformFn<YamlConfig.SnapshotTransformConfig> = async ({
  schema,
  config,
}): Promise<GraphQLSchema> => {
  // TODO: Needs to be changed!
  // eslint-disable-next-line no-eval
  const configIf = 'if' in config ? (typeof config.if === 'boolean' ? config.if : config.if && eval(config.if)) : true;

  if (configIf) {
    const resolvers = extractResolvers(schema);
    const resolversComposition: ResolversComposerMapping = {};

    const outputDir = isAbsolute(config.outputDir) ? config.outputDir : join(process.cwd(), config.outputDir);

    const snapshotComposition: ResolversComposition = next => async (root, args, context, info) => {
      const snapshotFilePath = computeSnapshotFilePath({
        typeName: info.parentType.name,
        fieldName: info.fieldName,
        args,
        outputDir,
      });
      if (await pathExists(snapshotFilePath)) {
        return readJSON(snapshotFilePath);
      }
      const result = await next(root, args, context, info);
      await writeFile(snapshotFilePath, result);
      return result;
    };

    for (const field of config.apply) {
      resolversComposition[field] = snapshotComposition;
    }

    const composedResolvers = composeResolvers(resolvers, resolversComposition);
    return addResolversToSchema({
      schema,
      resolvers: composedResolvers,
      updateResolversInPlace: true,
    });
  }

  return schema;
};

export default snapshotTransform;
