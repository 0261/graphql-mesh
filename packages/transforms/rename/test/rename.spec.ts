import { renameTransform } from './../src/index';
import { buildSchema, printSchema } from 'graphql';

describe('rename', () => {
  it('should change the name of a type', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Query {
        user: User!
      }

      type User {
        id: ID!
      }
    `);

    const newSchema = await renameTransform({
      schema,
      config: {
        type: '',
        from: 'User',
        to: 'MyUser'
      }
    });

    expect(newSchema.getType('User')).toBeUndefined();
    expect(newSchema.getType('MyUser')).toBeDefined();
    expect(printSchema(newSchema)).toMatchSnapshot();
  });
});
