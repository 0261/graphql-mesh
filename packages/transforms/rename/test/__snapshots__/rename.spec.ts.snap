// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rename should change the name of a type 1`] = `
"type MyUser {
  id: ID!
}

type Query {
  user: MyUser!
}
"
`;
